When an artifact is not found in a Maven repository, the plugin doesn't break the build

When a GAV is set to an artifact, that doesn't exist in a given Maven repository, p2-maven-plugin reports an exception in the log (see below), but the build is successful.

```
[INFO] --- p2-maven-plugin:1.3.0:site (default-cli) @ nightlab-pdfviewer-0.9.6-repo ---
[info] Resolved 0 features
[info] Resolving artifact=[org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200] transitive=[true] source=[true]
Downloading: http://an-repo/artifactory/repos/org/nightlab/org.nightlabs.base.ui/1.1.2.200810092200/org.nightlabs.base.ui-1.1.2.200810092200.pom
[WARNING] The POM for org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200 is missing, no dependency information available
Downloading: http://repository/artifactory/repos/org/nightlab/org.nightlabs.base.ui/1.1.2.200810092200/org.nightlabs.base.ui-1.1.2.200810092200.jar
[error] java.lang.RuntimeException: org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200 in public (http://repository/artifactory/repos/)
        at org.reficio.p2.P2Mojo.resolveArtifacts(P2Mojo.java:381)
        at org.reficio.p2.P2Mojo.processArtifacts(P2Mojo.java:288)
        at org.reficio.p2.P2Mojo.createFeature(P2Mojo.java:438)
        at org.reficio.p2.P2Mojo.processFeatures(P2Mojo.java:359)
        at org.reficio.p2.P2Mojo.execute(P2Mojo.java:244)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200 in public (http://repository/artifactory/repos/)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:384)
        at org.eclipse.aether.RepositorySystem$resolveDependencies$0.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
        at org.reficio.p2.resolver.maven.impl.AetherResolver.resolveWithTransitive(AetherResolver.groovy:106)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:210)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:174)
        at org.reficio.p2.resolver.maven.impl.AetherResolver.resolveBinaries(AetherResolver.groovy:84)
        at org.reficio.p2.resolver.maven.impl.AetherResolver.resolve(AetherResolver.groovy:61)
        at org.reficio.p2.P2Mojo.resolveArtifact(P2Mojo.java:416)
        at org.reficio.p2.P2Mojo.resolveArtifacts(P2Mojo.java:371)
        ... 25 more
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200 in public (http://repository/artifactory/repos/)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:459)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:262)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:367)
        ... 43 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.nightlab:org.nightlabs.base.ui:jar:1.1.2.200810092200 in public (http://repository/artifactory/repos/)
        at org.eclipse.aether.connector.wagon.WagonRepositoryConnector$6.wrap(WagonRepositoryConnector.java:1012)
        at org.eclipse.aether.connector.wagon.WagonRepositoryConnector$6.wrap(WagonRepositoryConnector.java:1004)
        at org.eclipse.aether.connector.wagon.WagonRepositoryConnector$GetTask.run(WagonRepositoryConnector.java:725)
        at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:67)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[INFO] Command line:
        [C:\devel\java\jdk1.8.0_181_x64\jre\bin\java.exe, -jar, C:\devel\.m2\repository\org\eclipse\tycho\tycho-bundles-external\1.0.0\eclipse\plugins\org.eclipse.equinox.launcher_1.3.201.v20161025-1711.jar, -configuration, C:\Users\sparsick\AppData\Local\Temp\tycho-p2-runtime8375830746209786104.tmp\configuration, -nosplash, -application, org.eclipse.equinox.p2.publisher.FeaturesAndBundlesPublisher, -artifactRepository, file:/C:/Users/sparsick/workspace/mes-nightlab-pdfviewer-repo/target/repository, -metadataRepository, file:/C:/Users/sparsick/workspace/mes-nightlab-pdfviewer-repo/target/repository, -publishArtifacts, -compress, -source, C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\target\source]
Generating metadata for ..
Generation completed with success [0 seconds].
[INFO] Command line:
        [C:\devel\java\jdk1.8.0_181_x64\jre\bin\java.exe, -jar, C:\devel\.m2\repository\org\eclipse\tycho\tycho-bundles-external\1.0.0\eclipse\plugins\org.eclipse.equinox.launcher_1.3.201.v20161025-1711.jar, -configuration, C:\Users\sparsick\AppData\Local\Temp\tycho-p2-runtime8072298705850024425.tmp\configuration, -nosplash, -application, org.eclipse.equinox.p2.publisher.CategoryPublisher, -categoryDefinition, file:/C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\target\repository\category.xml, -metadataRepository, file:/C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\target\repository]
Generating metadata for ..
Generation completed with success [0 seconds].
[INFO]
[INFO] --- tycho-p2-repository-plugin:0.25.0:archive-repository (default) @ nightlab-pdfviewer-0.9.6-repo ---
[INFO] Building zip: C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\target\nightlab-pdfviewer-0.9.6-repo-1-SNAPSHOT.zip
[INFO]
[INFO] --- build-helper-maven-plugin:3.0.0:attach-artifact (default) @ nightlab-pdfviewer-0.9.6-repo ---
[INFO]
[INFO] --- maven-install-plugin:2.4:install (default-install) @ nightlab-pdfviewer-0.9.6-repo ---
[INFO] Installing C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\pom.xml to C:\devel\.m2\repository\com\thyssenkrupp\tkse\mes\jersey-repo\nightlab-pdfviewer-0.9.6-repo\1-SNAPSHOT\nightlab-pdfviewer-0.9.6-repo-1-SNAPSHOT.pom
[INFO] Installing C:\Users\sparsick\workspace\mes-nightlab-pdfviewer-repo\target\nightlab-pdfviewer-0.9.6-repo-1-SNAPSHOT.zip to C:\devel\.m2\repository\com\thyssenkrupp\tkse\mes\jersey-repo\nightlab-pdfviewer-0.9.6-repo\1-SNAPSHOT\nightlab-pdfviewer-0.9.6-repo-1-SNAPSHOT.zip
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 9.041 s
[INFO] Finished at: 2019-10-11T09:11:33+02:00
[INFO] Final Memory: 30M/578M
[INFO] ------------------------------------------------------------------------
```
My expectation is if an artifact is not found, that p2-maven-plugin cancels the build because the resulted p2 repository doesn't make sense (see also below directory tree).

```
$ tree
.
├── artifacts.jar
├── category.xml
└── content.jar

0 directories, 3 files
```

--- a/README.md
+++ b/README.md
@@ -533,6 +533,9 @@ Example usage:
 	</configuration>
 ```
 
+Note: Up to version 1.3 unresolvable artifacts inside `<featureDefinitions>` did not cause the build to fail.
+The problem was only logged and the artifacts were omitted from the feature.
+This has been corrected in version 1.4 and the build will fail in such situations.
 
 ### Other features
 * p2-maven-plugin will tweak the version of a snapshot dependency replacing the SNAPSHOT string with a timestamp in the following format "yyyyMMddHHmmss" (feature #14)

--- a/src/main/java/org/reficio/p2/P2Mojo.java
+++ b/src/main/java/org/reficio/p2/P2Mojo.java
@@ -468,7 +468,7 @@ public class P2Mojo extends AbstractMojo implements Contextualizable {
 			log.info("Created feature "+p2featureDefinition.getId());
 
     	} catch (Exception e) {
-    		log.error(e);
+    		throw new RuntimeException(e);
     	}
     }
 

--- a/src/test/integration/bug-143-build-feature-non-existent-artifact-it/invoker.properties
+++ b/src/test/integration/bug-143-build-feature-non-existent-artifact-it/invoker.properties
@@ -1,0 +1,21 @@
+#
+# Copyright (c) 2012 Reficio (TM) - Reestablish your software! All Rights Reserved.
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+invoker.goals=p2:site
+invoker.buildResult = failure

--- a/src/test/integration/bug-143-build-feature-non-existent-artifact-it/pom.xml
+++ b/src/test/integration/bug-143-build-feature-non-existent-artifact-it/pom.xml
@@ -1,0 +1,70 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (C) 2006-2007 the original author or authors.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+-->
+
+<!-- $Id$ -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.reficio</groupId>
+        <artifactId>integration</artifactId>
+        <version>@project.version@</version>
+        <relativePath>../integration.xml</relativePath>
+    </parent>
+
+    <artifactId>config</artifactId>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+    </properties>
+
+    <description>
+        Test building feature should fail when non-existent artifact is referenced and skipNotExistingArtifacts option is not set
+    </description>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.reficio</groupId>
+                <artifactId>p2-maven-plugin</artifactId>
+                <version>@project.version@</version>
+                <executions>
+                    <execution>
+                        <id>default-cli</id>
+                        <configuration>
+                            <featureDefinitions>
+                                <feature>
+                                    <id>org.foo.bar.feature</id>
+                                    <version>1.0.0</version>
+                                    <artifacts>
+                                        <artifact>
+                                            <id>org.foo.bar:nonexistent:1.0.0</id>
+                                        </artifact>
+                                    </artifacts>
+                                </feature>
+                            </featureDefinitions>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>

